(in progress) preserve HTML block and br elements in GFM tables

jywarren opened the pull request.

jywarren committed:
initial test for block level element preservation in GFM tables

initial inTable state

preserving block elements in table cells, spacing issues remain

blockTags and <br> vs <br /> simplification

finalized br handling and HTML within pre blocks with test

added <mark> in <table> test

jywarren commented:
This is pretty close; three things remain:
1. I noticed that html tags inside of `<code>` blocks are also not preserved, and added a test for that too, since it's similar. But I could break that out into a distinct PR or issue if you prefer.
2. Spacing: I'm assuming that we want the table formatting to match the width of the content even with lots of HTML tags. I'll take a look at adjusting that soon too; it's reflected in the test. 
3. `<br>` formatting. I'll leave an inline comment.

jywarren commented on domador.js:
```diff
@@ -150,7 +150,8 @@ Domador.prototype.append = function append (text) {
 Domador.prototype.br = function br () {
-  this.append('  ' +  this.left);
+  if (!this.inTable) this.append('  ' +  this.left);
+  else this.append('  <br />');
```
I started here, and it results in cleaner `<br />`s, but I have an alternative version which does not add these two lines, and processes `<br />`s the same way as other block elements. That version is simpler and single-track, but won't clean up a messy `bar<br>` into `bar <br />` the way this will. Preference?

bevacqua replied:
I'd rather do all of them in a single place. Not concerned about `<br>` vs `<br/>`.

jywarren commented on domador.js:
```diff
@@ -370,12 +371,17 @@ Domador.prototype.process = function process (el) {
   }
 
+  if (this.inTable) console.log("INTABLE ", el.tagName);
+  if (this.inTable && ['DIV', 'P', 'UL', 'LI', 'PRE', 'BLOCKQUOTE'].indexOf(el.tagName) !== -1) {
```
I alphabetized this in the most recent, don't worry. And any other block elements I should add here?

bevacqua replied:
Use all of `paragraphTags` plus `'P', 'UL', 'LI', 'PRE'` here. Maybe name it `blockTags` below `paragraphTags`.

jywarren commented:
Oh one more question before I close my laptop :-) -- is column width set by header width or by largest cell width? I'm trying to follow the code but knowing ahead of time will make it easier for me to parse through this, thanks.

jywarren commented:
OK, I think I've addressed the issues above, and also integrated preserving HTML within `<pre>` blocks, with a couple tests for that as well. How does this look?

bevacqua commented on domador.js:
```diff
@@ -41,6 +41,9 @@ var shallowTags = [
   'ADDRESS', 'ARTICLE', 'ASIDE', 'DIV', 'FIELDSET', 'FOOTER', 'HEADER', 'NAV', 'P', 'SECTION'
 ];
+var blockTags = [
+  'ADDRESS', 'ARTICLE', 'ASIDE', 'DIV', 'FIELDSET', 'FOOTER', 'HEADER', 'NAV', 'P', 'SECTION', 'UL', 'LI', 'BLOCKQUOTE', 'BR'
```
TABLE TR TD?

jywarren replied:
Nested tables? I guess there's no reason not to; I can give it a try and if it causes unexpected issues people can file them.

bevacqua commented:
Looking pretty good. Could you add a few tests like the ones for `<mark>` tags? I'm pretty sure this breaks `<mark>` in `<table>` and `<pre>`

jywarren commented:
Sure, i'll give it a go and rebase-squash this all.

jywarren commented:
Actually I'm not finding documentation for the `<mark>` tag -- it's not in commonmark... where could I find a description of its usage or spec?

jywarren commented:
Oh, sorry, it's HTML, not markdown specific! Got it.

jywarren commented:
So, just to confirm, is the `<mark>` behavior that it (and no other element) is preserved, and can contain markdown? And specifically what interaction do you want to test between `<mark>` and `<table>` or `<pre>` -- that tags inside of `<mark>` and then inside of `<table>` would still be converted to markdown? Like:

```
<table>
  <tr>
    <th>col0</th>
  </tr>
  <tr>
    <td><mark><em>foo</em></mark></td>
  </tr>
</table>
```

becomes:

```
| col0 |
|------|
| <mark> _foo_ </mark> |
```

Is that right? Thanks!

bevacqua commented:
Yup! Maybe copy over the `<mark>` test cases and see whether they pass if the markup is wrapped with a table.

jywarren commented:
Added a test -- it actually seems to work (as described in my last comment). Also rebased and squahed. Another test?

bevacqua commented:
Looks almost good to go, seems we're missing the nested raw `table tr td`

jywarren commented:
Sorry, doing too many things at once, as usual. Do we also need to do `thead tbody`?

I tried adding `table tr td` and unfortunately it breaks a lot of tests. Do you think we could address that separately? I'd be happy to open an issue and mark line numbers etc, and it's not likely to be a common case (nesting tables in tables).

bevacqua commented:
Yes to all. Go ahead and merge this!

jywarren commented:
Do you have a preference for merge vs squash and merge, in the newish GitHub options here?

bevacqua commented:
Merge is okay. You already squashed

jywarren commented:
Done! Thanks!

jywarren commented:
Hi, would you mind making a new release for this? I'm going to pull it into https://github.com/publiclab/PublicLab.Editor via `woofmark`. Thanks!

bevacqua commented:
Done.
